connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/outbound/JobHandlerContext.java

```diff
@@ -35,29 +36,54 @@ public class JobHandlerContext extends AbstractConnectorContext
 
   private final ActivatedJob job;
 
-  public JobHandlerContext(final ActivatedJob job, final SecretProvider secretProvider) {
-    super(secretProvider);
+  private final ObjectMapper objectMapper;
+
+  private String jsonWithSecrets = null;
+
+  public JobHandlerContext(
+      final ActivatedJob job,
+      final SecretProvider secretProvider,
+      final ValidationProvider validationProvider,
+      final ObjectMapper objectMapper) {
+    super(secretProvider, validationProvider);
     this.job = job;
+    this.objectMapper = objectMapper;
   }
 
   @Override
-  public <T> T getVariablesAsType(Class<T> cls) {
-    return job.getVariablesAsType(cls);
+  public Map<String, String> getCustomerHeaders() {
+    return job.getCustomHeaders();
   }
 
   @Override
-  public String getVariables() {
-    return job.getVariables();
+  public <T> T bindVariables(Class<T> cls) {
+    var mappedObject = mapJson(getJsonReplacedWithSecrets(), cls);
+    getValidationProvider().validate(mappedObject);
+    return mappedObject;
   }
 
-  @Override
-  public Map<String, String> getCustomHeaders() {
-    return Optional.ofNullable(job.getCustomHeaders()).orElse(Collections.emptyMap());
+  private String getJsonReplacedWithSecrets() {
+    if (jsonWithSecrets == null) {
+      try {
+        jsonWithSecrets = getSecretHandler().replaceSecrets(job.getVariables());
+      } catch (Exception e) {
+        throw new ConnectorException("SECRET_MAPPING", "Error during secret mapping.");
```

Comment:
nit: Should we include the message at least in some cases? e.g. if it's a not found error, I would appreciate a hint which secret I forgot to add in the console.

Reply:
@chillleader Yes it would but that information is afaik not provided by the SDK as of now as `null` is a supported return value from a `SecretProvider` implementation.

Reply:
I could imagine that we return a result as part of the `replaceSecrets` operation that contains the mapped result but also reports maybe on missing values along the way that we can then make use of here. I could see this as an additional improvement and will add it to the architecture session.

Reply:
@chillleader Agree, would you be ok to throw a dedicated `SecretException` in `https://github.com/camunda/connector-sdk/blob/f7a2d15ce83acd6d9140611d7a9ab4ecd6ab4dfa/core/src/main/java/io/camunda/connector/impl/secret/SecretUtil.java#L68` in your PR and map this exception here in your PR as well?  As this requires an SDK update as well it might be leaner this way. WDYT?

Reply:
Works for me